Makefile

```diff
@@ -250,22 +266,37 @@ depend.update.full: depend.cleanlock depend.update
 depend.cleanlock:
 	-rm Gopkg.lock
 
-depend.update:
+depend.update: | $(DEP)
 	@echo "Running dep ensure with DEPARGS=$(DEPARGS)"
-	time dep ensure $(DEPARGS)
+	time $(DEP) ensure $(DEPARGS)
 	cp Gopkg.* vendor/
 	@echo "now check the diff in vendor/ and make a PR"
 
-vendor.check:
-	@echo "Checking that Gopkg.* are in sync with vendor/ submodule:"
-	@echo "if this fails, 'make pull' and/or seek on-call help"
-	diff Gopkg.toml vendor/
-	diff Gopkg.lock vendor/
-
 .PHONY: vendor.check
+vendor.check: $(ISTIO_OUT)/vendor_checked
+
+$(ISTIO_OUT)/vendor_checked: Gopkg.toml Gopkg.lock | $(ISTIO_OUT) $(GRPC_DIR)
+	@diff Gopkg.toml vendor/ || (echo "vendor/ out-of-sync, run 'make pull' and/or seek on-call help" && exit 1)
+	@diff Gopkg.lock vendor/ || (echo "vendor/ out-of-sync, run 'make pull' and/or seek on-call help" && exit 2)
+	touch $(ISTIO_OUT)/vendor_checked
 
 ${GEN_CERT}:
-	unset GOOS && unset GOARCH && CGO_ENABLED=1 bin/gobuild.sh $@ istio.io/istio/pkg/version ./security/cmd/generate_cert
+	GOOS=$(LOCAL_GOOS) GOARCH=$(LOCAL_GOARCH) bin/gobuild.sh $@ istio.io/istio/pkg/version ./security/cmd/generate_cert
+
+# If CGO_ENABLED=0 then go get tries to install in system directories.
+# If -pkgdir <dir> is also used then various additional .a files are present.
+${ISTIO_BIN}/dep:
```

Comment:
where/how do we make sure that GOPATH/bin is (first) in the PATH ?
(ie that installing dep and fortio will be using the one just installed)

Reply:
There's been some portion of istio developers who feel that if they have certain golang tools installed at the system level (or preexisting elsewhere in the path) then the build should use/honor those (even if it means the build loses control over which version is used).  I'm not quite in agreement with the tradeoffs of that philosophy but have tried to accommodate it nonetheless.

Using dep as an example (which also applies analogously to fortio and helm in this change), the "dep" that's used is decided by:
<pre>
DEP       := $(shell which dep    2>/dev/null || echo "${ISTIO_BIN}/dep" )
</pre>
I.e., favor any preinstalled dep, otherwise use one at ${ISTIO_BIN}/dep.  The line where you commented tells make how to create ${ISTIO_BIN}/dep if necessary (it's intentional that the target doesn't use the simpler "${DEP}:" since that macro could be pointing elsewhere if dep is pre-installed in the path, and "go get" isn't likely to populate that location).

The final related part is that any make target that has a dependency on ${DEP} does so as an order-only dependency.  If ${DEP} points elsewhere and a normal dependency is used then make would complain that it doesn't know how to make ${DEP}, so an order-only means that make only cares that the file exists (and make doesn't care that it doesn't know how to create the file unless/until the file is missing).  The tradeoff to this is that an existing ${ISTIO_BIN}/dep will never get updated.  So, there's an expectation/reliance that these tools remain functional/compatible over time (otherwise a dev needs to clean out "${ISTIO_BIN}/").  Helm doesn't suffer from this particular staleness issue since the version is part of the filename ("${ISTIO_BIN}/helm$(HELM_VER)" ) so if/when HELM_VER gets bumped then make will see the file is not present and try to download a new one.

If staleness is a concern that I suppose a version could be added to the names of dep and fortio, e.g.:

<pre>
DEP_VER := 1 # just an arbitrary value someone changes when they want to force an update
DEP       := $(shell which dep    2>/dev/null || echo "${ISTIO_BIN}/dep${DEP_VER}" )
${ISTIO_BIN}/dep${DEP_VER}:
        unset GOOS && unset GOARCH && CGO_ENABLED=1 go get -u github.com/golang/dep/cmd/dep <b>&& mv ${ISTIO_BIN}/dep ${ISTIO_BIN}/dep${DEP_VER}</b>
</pre>